compiler/ast/ast_query.nim

```diff
@@ -690,6 +690,16 @@ iterator genericParamsInMacroCall*(macroSym: PSym, call: PNode): (PSym, PNode) =
     if posInCall < call.len:
       yield (genericParam, call[posInCall])
 
+proc endsInNoReturn*(n: PNode): bool =
+  ## checks if expression `n` ends in an unstructured exit (raise, return,
+  ## etc) or calls of noreturn proc. This is meant to be called on an semmed `n`.
+  var it = n
+  while it.kind in {nkStmtList, nkStmtListExpr} and it.len > 0:
+    it = it.lastSon
+  result = it.kind in nkLastBlockStmts or
+    it.kind in {nkIfStmt, nkIfExpr, nkTryStmt, nkCaseStmt, nkTryStmt, nkBlockExpr} and it.typ.isNil or
```

Comment:
```suggestion
    (it.kind in {nkIfStmt, nkTryStmt, nkCaseStmt, nkBlockStmt} and
     it.typ.isEmptyType()) or
```
`nkIfExpr` and `nkBlockExpr` don't need to be handled here, since they're guaranteed to have a non-`void` type.

Some noreturn compound statements can also have a non-nil `typ` (when assigned the `enforceVoidContext` type), so `isEmptyType` (which returns true for both `typ == nil` and `typ.kind == tyVoid`) has to be used.